Update Badging article for stable version. (#2321)
GoogleChrome/web.dev: src/site/content/en/blog/badging-api/index.md

  - petelepage
description: The Badging API allows installed web apps to set an application-wide badge, shown in an operating-system-specific place associated with the application, such as the shelf or home screen. Badging makes it easy to subtly notify the user that there is some new activity that might require their attention, or it can be used to indicate a small amount of information, such as an unread count.
date: 2018-12-11

> updated: 2020-03-16

Author: ```suggestion
updated: 2020-03-13
```

  - origin-trial
hero: hero.jpg
alt: Phone showing several notification badges
---

{% Aside 'success' %}
  This API is part of the new

>   [capabilities project](https://web.dev/fugu-status/),

Reviewer: ```suggestion
  [capabilities project](/fugu-status/),
```

Since it's on the same domain, you don't need the FQDN

  and it is available in Chrome by default starting in Chrome 82.
{% endAside %}

## What is the Badging API? {: #what }

> <figure class="w-figure w-figure--center">

Author: As it was, this figure is very hard to make out.

| ------------------------------------------ | ---------------------------- |
| 1. Create explainer                        | [Complete][explainer]        |
| 2. Create initial draft of specification   | [Complete][spec]             |
| 3. Gather feedback & iterate on design     | Complete                     |
| 4. Origin trial                            | Complete                     |

> | **5. Launch**                              | Complete                     |

Reviewer: ```suggestion
| **5. Launch**                              | Chrome 82                    |
```

> | **5. Launch**                              | Chrome 82                    |

Reviewer 2: It's going out in 81.

Reviewer 3: Badging first landed in 82, and was later merged into 81. Maybe that's why 82 is mentioned. So yes, badging is enabled in 81 and is not experimental.

</div>

## Try it

> 1. Using Chrome 82 or later on Windows or Mac, open the [Badging API demo][demo].

Reviewer 2: 81

## How to use the Badging API {: #use }

The Badging API works on Windows, and macOS. Support for Chrome OS is in
development and will be available in a future release of Chrome.
On Android, the Badging API is not supported. Instead, Android automatically

> shows a badge on a web app when there is an unread notification, just as

Reviewer: "just as for native apps" reads oddly to me, would "the same as native apps" be better?

// Clear the badge
navigator.clearAppBadge();
```

You can only call `setAppBadge()` and `clearAppBadge()`
from a foreground page, or potentially in the future, a service worker. In
either case, it affects the whole app, not just the current page.

In some cases, the OS may not allow the exact representation of the badge.
In such cases, the browser will attempt to provide the best representation for
that device. For example, because the Badging API isn't supported on Android,
Android only ever shows a dot instead of a numeric value.

Don't assume anything about how the user agent displays the badge.
Some user agents may take a number like "4000" and rewrite it as
"99+". If you saturate the badge yourself (for example by setting it to "99")
then the "+" won't appear. No matter the actual number, just call
`setAppBadge(unreadCount)` and let the user agent deal with
displaying it accordingly.

While the Badging API *in Chrome* requires an installed app, you shouldn't
make calls to the Badging API dependent on the install state. Just call the
API when it exists, as other browsers may show the badge in other places.
If it works, it works. If not, it simply doesn't.

> 

Reviewer 2: I think it's premature to remove the "Supporting both the first and second origin trial" sections, since a similar thing applies here. Perhaps remove the first origin trial since that's long gone. "Supporting both the origin trial and released versions".

Remove the else if clauses for "window.ExperimentalBadge" since that's the first origin trial. And instead, do "if (navigator.setAppBadge) { ... } else if (navigator.setExperimentalAppBadge) { ... }".

Reviewer 2: We're also still happy to hear feedback about the API design, though we don't want to make major incompatible changes, we can still add functionality.

Reviewer: @jpmedley - can we add the section on supporting both back then please?

Author: @mgiuca Something's not making sense to me. OT tokens are deactivated when the OT ends. Are you telling me that tokens for the second trial are still being honored?

Reviewer 2: Yes, tokens from the second trial are still being honored _on Chrome 80 and earlier_. Sites will want to continue working on Chrome 80 which still uses the `setExperimentalAppBadge` API, and this is a way to let them do that.

Author: Readding the feedback section.

Pardon me for being a pest, it seems lie you wouldn't want to issue new tokens at this point.

Reviewer: If a user doesn't upgrade and stays on Chrome 80, the existing token will continue to function for some period of time.

Author: If that's your suggested rewrite, I agree.

### Report a problem with the implementation

Did you find a bug with Chrome's implementation? Or is the implementation
different from the spec?

* File a bug at <https://new.crbug.com>. Be sure to include as much detail as
  you can, simple instructions for reproducing, and set **Components** to
  `UI>Browser>WebAppInstalls`. [Glitch](https://glitch.com) works great for
  sharing quick and easy reproductions.

### Show support for the API

> Planning to use the Badging API on your site? Your public support helps the

Reviewer 2: Policy thing: do we want to keep this section now that the API is launching? Is this a thing we ask people to do for all APIs, or just ones in a trial?

Reviewer: I think it's good to still ask for support, it helps other browser vendors see people are using it, and it also potentially points us to new sites that are using it that we can point people at.

The Badge API consists of the following methods on `navigator`:

* `setAppBadge([number])`: Sets the app's badge. If a value is
  provided, set the badge to the provided value otherwise, display a plain
  white dot (or other flag as appropriate to the platform).
* `clearAppBadge()`: Removes app's badge.

{% Aside %}
These functions had different names for the origin trial than they do in the

> current spec. If you're still participating in the origin trial, you'll need to

Reviewer: ```suggestion
current implementation. If you're still participating in the origin trial, you'll need to
```

support both versions of the API. See [Supporting both the origin trial and the
stable API](#supporting) for details.
{% endAside %}

For example:

```js
// Set the badge
const unreadCount = 24;
navigator.setAppBadge(unreadCount);

// Clear the badge
navigator.clearAppBadge();
```

You can only call `setAppBadge()` and `clearAppBadge()`

> from a foreground page, or potentially in the future, a service worker. In

Reviewer: ```suggestion
from a foreground page. In the future, you may be able to call it from a service worker. In
```

either case, it affects the whole app, not just the current page.

In some cases, the OS may not allow the exact representation of the badge.
In such cases, the browser will attempt to provide the best representation for
that device. For example, because the Badging API isn't supported on Android,
Android only ever shows a dot instead of a numeric value.

Don't assume anything about how the user agent displays the badge.
Some user agents may take a number like "4000" and rewrite it as
"99+". If you saturate the badge yourself (for example by setting it to "99")
then the "+" won't appear. No matter the actual number, just call
`setAppBadge(unreadCount)` and let the user agent deal with
displaying it accordingly.

While the Badging API *in Chrome* requires an installed app, you shouldn't
make calls to the Badging API dependent on the install state. Just call the
API when it exists, as other browsers may show the badge in other places.
If it works, it works. If not, it simply doesn't.

## Supporting both the origin trial and the stable API {: #supporting }

In the Chrome 79 and Chrome 80 [origin trial]({{origin_trial.url}}) the members

> of the badging API have different names than those in the current spec. If

Reviewer: ```suggestion
of the badging API have different names than those in the current implementation. If
```

you're participating in that origin trial, you'll need to support both the
origin trail and the standard names. You can wrap the calls in the following
functions. Once the origin trial is over or all your users have migrated to
Chrome 81 or later, you can remove these functions from your code.

```js
function setBadge(...args) {
  if (navigator.setExperimentalAppBadge) {
    navigator.setExperimentalAppBadge(...args);

>   } else {

Reviewer 2: Nit: I would prefer to swap these cases around, so the "normal" case is the final API, with a fallback to the experimental API. (That's how it was structured previously, check for the new API first, fall back to the old API.)

So `if (navigator.setAppBadge) { do it } else { old one }`.

Author: I don't remember why I did it that way. @petele why did you approve it that way?

    window.setAppBadge(...args);
  }
}

function clearBadge() {
  if (navigator.clearExperimentalAppBadge) {
    navigator.clearExperimentalAppBadge();
  } else {

>     window.clearAppBadtge(...args);

Reviewer: ```suggestion
    window.clearAppBadge(...args);
```
